XML refactors: pcmk__message_add_xml() to mark_xml_changes()

nrwahl2 opened the pull request:
Next batch for after #3397 is agreed-upon and merged

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -1147,7 +1071,7 @@ pcmk__xml_needs_escape(const char *text, enum pcmk__xml_escape_type type)
         }
 
-        text++;
+        text = g_utf8_next_char(text);
```
We could alternatively use `xmlGetUTF8Size()`. I've considered that, since the purpose of all this is to escape text that will be part of an XML tree, and one could argue that we should use libxml2's logic to the extent possible.

I used `g_utf8_next_char()` for a few reasons:
1. I found it first
2. It performs better because it doesn't validate the encoding
3. It doesn't require a temporary "length" output variable
4. In general (knock on wood) GLib seems to be less buggy than libxml2.

In an upcoming PR I use `xmlGetUTF8Char()` (not to be confused with `xmlGetUTF8Size()`) in an updated version of `crm_xml_sanitize_id()`.

kgaillot commented on lib/common/xml.c:
```diff
@@ -1001,74 +1001,6 @@ crm_xml_set_id(xmlNode *xml, const char *format, ...)
- * \return Number of non-ASCII UTF-8 bytes at the beginning of \p text
- */
-static size_t
-utf8_bytes(const char *text)
```
If you want to be paranoid, require CHAR_BIT to be 8 in configure.ac

nrwahl2 replied:
good idea

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -460,29 +460,47 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
+        if (xml_acl_denied(target)) {
+            crm_trace("Cannot copy %s=%s to %s",
+                      name, value, (const char *) target->name);
+            return EPERM;
```
Hmm, I was going by this:
https://www.gnu.org/software/libc/manual/html_node/Error-Codes.html

> Macro: int EPERM
> "Operation not permitted." Only the owner of the file (or other resource) or processes with special privileges can perform the operation.
> ...
> Macro: int EACCES
> "Permission denied." The file permissions do not allow the attempted operation.

Then there's the POSIX description, which I just now found:
https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_03

> [EACCES]
> Permission denied. An attempt was made to access a file in a way forbidden by its file access permissions.
> ...
> [EPERM]
> Operation not permitted. An attempt was made to perform an operation limited to processes with appropriate privileges or to the owner of a file or other resource.

To me, it's not clear-cut.
* We're not accessing a file; we're performing an operation that requires particular privileges. That suggests `EPERM`.
* I can see an argument in favor of `EACCES` based on "[access] permissions."

kgaillot replied:
Huh, not sure where I saw the other definitions. Either one is close enough for me.

nrwahl2 commented on lib/common/crmcommon_private.h:
```diff
@@ -135,6 +135,11 @@ bool pcmk__marked_as_deleted(xmlAttrPtr a, void *user_data);
 
+#if defined(PCMK__UNIT_TESTING)
+G_GNUC_INTERNAL
+void pcmk__xa_set_expand(xmlNode *target, const char *name, const char *value);
```
> I'd rather just make things library-private than special-case unit testing

That's fine. I don't *like* it, because it sort of violates the spirit of things... I prefer keeping scopes as tight as possible. And it means that in an ideal world where everything is unit-tested, we'd have few or no static functions.

However, in the absence of a way to unit-test a static function, making them library-private is probably the lesser evil. The `defined` guards are clunky.

(Though if we really wanted, it might be possible to make a macro with a short name that expands to these CPP directives. I think clumens and I tossed around something like `PCMK__STATIC` a while back. Not sure about that kind of CPP trickery off the top of my head... it would need to expand and then get processed again.)

> `pcmk__xe_set_score()`

that's probably good

kgaillot replied:
Yeah, I considered a macro as well, but as a general principle it's a good idea to avoid any differences between unit testing and actual calls. Not always possible, but it's best to minimize it.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -460,14 +460,91 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
+     *
+     * @COMPAT Should we try to expand, assuming 0 for the old value? This
+     * doesn't seem fundamentally different from the "expanding ourselves" case
+     * mentioned below -- it's old=<unset>/new="X++" versus old="X++"/new="X++".
```
I'm hoping I can make `fix_plus_plus_recursive()` go away. There's only one caller. I've been putting it off because it means tracing through more convoluted-as-hell CIB code than I'd like. But last time I looked at it, it seemed like it was impossible for it to actually do what it was put there to do.

kgaillot replied:
I believe I traced through it at one point, but I don't remember anything about it.

There was a time before attrd when attributes were always set directly in the CIB, and it may date to then. It may also be usable with `CIB_file`.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -460,14 +460,91 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
+        const char *v = value;
+
+        // Stop at first character that differs between name and value
+        for (; *n == *v; n++, v++);
```
Ouch, good catch.

nrwahl2 commented on daemons/attrd/attrd_utils.c:
```diff
@@ -305,7 +305,7 @@ void
 {
     /* Copy attributes from the wrapper parent node into the child node.
-     * We can't just use copy_in_properties because we want to skip any
+     * We can't just use pcmk__xe_copy_attrs because we want to skip any
```
good point

nrwahl2 replied:
also would need to account for "++"/"+=" expansion -- this function currently ignores them, but `pcmk__xe_copy_attrs()` expands them. I don't want to introduce any new expansions when we're planning to deprecate them.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -460,14 +460,91 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
+            /* value="X++": new value of X is old_value + 1
+             * value="X+=Y": new value of X is old_value + Y (for some number Y)
+             */
+            int add = (*v == '+')? 1 : char2score(++v);
```
> which also means that updating "foo" to "foo+=bar" will result in "0"

It would result in `foo = foo`, ultimately. This is just determining what we add to `foo`

nrwahl2 replied:
if `foo` originally had a score in it; maybe you're talking about if it doesn't

kgaillot replied:
Right, if an attribute "foo" is initially set to the literal value "foo", `char2score()` will treat that as 0, and if it's set to the literal value "foo+=bar", `char2score()` will treat "bar" as 0, and the attribute will end up being set to 0

clumens commented on lib/pengine/bundle.c:
```diff
@@ -1559,23 +1559,23 @@ pe__bundle_xml(pcmk__output_t *out, va_list args)
         if (print_ip) {
-            out->message(out, crm_map_element_name(replica->ip->xml), show_opts,
-                         replica->ip, only_node, only_rsc);
+            out->message(out, pcmk__map_element_name(replica->ip->xml),
```
We talked about this in T787 some, and removing the calls won't be necessary.  Everything will be handled by schema transforms.  So, I'm going to leave them alone.

kgaillot replied:
@clumens is just deprecating the function for 2.1.8, which if you're doing that here, takes care of that. For 3.0.0, we'll drop the functions and all calls everywhere.

My point for a bundle's implicit resources is that we define them, as primitives not clones, so we never needed the function in those places. (`bundle->child` does need it for now however.)

nrwahl2 replied:
Thanks, makes sense.

> it's always primitive, but we could use `->name` for future-proofing

Hmm I like `PCMK__XE_PRIMITIVE` where we know that to be the case, but `->name` makes explicit where the message name is coming from. I lean toward `->name` for that reason, on top of future proofing.

nrwahl2 commented on include/crm/common/messages_internal.h:
```diff
@@ -119,4 +120,28 @@ pcmk__request_origin(const pcmk__request_t *request)
+{
+    // No way to retrieve the copy later if msg == NULL
+    CRM_CHECK(msg != NULL, return);
+    pcmk__xml_copy(pcmk__xe_create(msg, name), src);
```
> obscures what's being done

Kind of yeah. It looks a little odd when expanded FWIW, with the payload inside two layers of nesting. But that might still be better.

> rather than message-specific

yeah that always got me

> If we were starting from scratch

Yeah I think I would prefer that. I'm not looking forward to making that change though, if I'm ever the person to do it. Especially with rolling-upgrade compatibility in `pacemaker-based`.

nrwahl2 commented:
Addressed review of get/add message functions and pulled in commits from #3413. About to rebase on main.

Leaving in draft, lots more review to address

nrwahl2 commented on include/crm/common/xml_internal.h:
```diff
@@ -558,4 +559,17 @@ void pcmk__sort_schemas(void);
 // @COMPAT Remove when v1 patchsets are removed
 xmlNode *pcmk__diff_v1_xml_object(xmlNode *left, xmlNode *right, bool suppress);
 
+// @COMPAT Drop when "master" is removed
```
Thanks

nrwahl2 commented on lib/pengine/bundle.c:
```diff
@@ -1559,23 +1559,23 @@ pe__bundle_xml(pcmk__output_t *out, va_list args)
         if (print_child) {
-            out->message(out, crm_map_element_name(replica->child->xml), show_opts,
-                         replica->child, only_node, only_rsc);
+            out->message(out, pcmk__map_element_name(replica->child->xml),
```
This is a clone instance anyway, not the clone, so it's a primitive

kgaillot replied:
Ah right, bundles are always confusing with rsc->children vs bundle_data->child vs replica->child

kgaillot replied:
If I ever get around to best practices for that code, I think I'll rename bundle_data->child to bundle_data->clone and replica->child to replica->clone_instance

nrwahl2 replied:
That would be wonderful. I had a fair amount of trouble with those names when doing the "re-shuffling clone instances" work.

nrwahl2 commented on lib/pengine/bundle.c:
```diff
@@ -1703,23 +1703,27 @@ pe__bundle_html(pcmk__output_t *out, va_list args)
             if (print_child) {
-                out->message(out, crm_map_element_name(replica->child->xml),
-                             new_show_opts, replica->child, only_node, only_rsc);
+                out->message(out, pcmk__map_element_name(replica->child->xml),
```
the clone child is a primitive

nrwahl2 commented on lib/pengine/bundle.c:
```diff
@@ -1841,23 +1845,27 @@ pe__bundle_text(pcmk__output_t *out, va_list args)
             if (print_child) {
-                out->message(out, crm_map_element_name(replica->child->xml),
-                             new_show_opts, replica->child, only_node, only_rsc);
+                out->message(out, pcmk__map_element_name(replica->child->xml),
```
clone child is primitive

nrwahl2 commented on lib/pengine/group.c:
```diff
@@ -385,8 +385,8 @@ pe__group_xml(pcmk__output_t *out, va_list args)
 
-        out->message(out, crm_map_element_name(child_rsc->xml), show_opts, child_rsc,
-                     only_node, only_rsc);
+        out->message(out, pcmk__map_element_name(child_rsc->xml), show_opts,
```
oh good point

nrwahl2 replied:
We can also drop the ones for children in `clone.c`. We can't have cloned clones AFAIK.

nrwahl2 commented:
Next push addresses most of review. After this, I need to revisit the attribute expansion stuff.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -460,29 +460,47 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
-                return;
-            }
+        expand_plus_plus(target, name, value);
+        if (xml_acl_denied(target)) {
```
@kgaillot I think we can drop this check, as long as we give `copy_in_properties()` its own body so that the behavior is preserved there. `crm_xml_add()` checks ACLs. Between that and whatever final checking happens in `cib_perform_op()`, we should be fine. Let me know if you disagree.

nrwahl2 replied:
That would make it easier to reason about `pcmk__xe_copy_attrs()`, and would leave it with only one "extra" side effect: `++` expansion

kgaillot replied:
It seems OK as long as `update_cib_object()` still checks `xml_acl_denied()` after calling this.

nrwahl2 replied:
Thanks. The way I look at it, in places where we have
```
crm_xml_add(xml, name1, value1);
crm_xml_add(xml, name2, value2);
```

We don't call `xml_acl_denied()` in between those calls to determine whether to proceed. So it doesn't make much sense to do it when looping over a bunch of attrs, except for a tiny efficiency boost.

nrwahl2 replied:
`update_cib_object()` shouldn't need to check `xml_acl_denied()` either. Its only caller is `add_cib_object()`, whose only caller is `cib_process_create()`, which always goes through `cib_perform_op()`.

nrwahl2 replied:
Nevermind, I see what you mean, I forgot that it used to check `xml_acl_denied()` for setting its return value.

kgaillot replied:
> Thanks. The way I look at it, in places where we have
> 
> ```
> crm_xml_add(xml, name1, value1);
> crm_xml_add(xml, name2, value2);
> ```
> 
> We don't call `xml_acl_denied()` in between those calls to determine whether to proceed. So it doesn't make much sense to do it when looping over a bunch of attrs, except for a tiny efficiency boost if there is a denial.

ACLs are element-level, so the first attribute checked will be definitive.

nrwahl2 replied:
Right. We *can* check and short-circuit on ACL denial. What I'm saying is that it seems unnecessary, and that we *don't* do that in other contexts where we set multiple attributes for an element.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -460,14 +460,91 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
+            && ((*v == '+') || (*v == '='))) {
+
+            // If we're expanding ourselves, no previous value was set; use 0
+            int old_value_i = (old_value != value)? char2score(old_value) : 0;
```
I'm going to deprecate this function now as part of deprecating the "++"/"+=" expansion, so it's even less worth stepping in.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -1746,11 +1752,14 @@ pcmk__xml_update(xmlNode *parent, xmlNode *target, xmlNode *update,
-        /* No need for expand_plus_plus(), just raw speed */
+        /* No need for pcmk__xa_set_expand(), so use faster method
+         *
+         * @TODO pcmk__xe_copy_attrs() also checks ACLs. Should we do that here?
```
> Probably even less worth it, but if the extra copy_attrs argument were a flag group, it could cover this case too

I'll ignore it since this block will go away when `apply_xml_diff()` goes away (hopefully in 3.0.0, as it's been deprecated since 2.0.5).

nrwahl2 commented:
This got bigger than expected.
* Deprecating the "++"/"+=" syntax led to a small detour into libcib. `update_cib_object()` was a hot mess.
* I added more unit tests for `pcmk__xe_copy_attrs()`. I spent a while debugging an abort, which turned out to be because we don't in general call `crm_xml_init()` before unit tests that manipulate XML. A `doc->_private` was `NULL`, causing an assertion.

nrwahl2 commented:
no code changes, just added "Closes T808" to a commit message

kgaillot reviewed (commented):
reviewed through "Drop update_cib_object()"

nrwahl2 commented:
> reviewed through "Drop update_cib_object()"

Most of the rest hasn't changed since the previous push FYI

nrwahl2 commented on daemons/based/based_callbacks.c:
```diff
@@ -1317,13 +1317,16 @@ prepare_input(const xmlNode *request, enum cib__op_type type,
+                input = pcmk__xe_first_child(wrapper, NULL, NULL, NULL);
             }
-            wrapper = pcmk__xe_first_child(request, wrapper_name, NULL, NULL);
-            input = pcmk__xe_first_child(wrapper, NULL, NULL, NULL);
```
Thanks

nrwahl2 commented on lib/cib/cib_ops.c:
```diff
@@ -670,22 +649,23 @@ add_cib_object(xmlNode * parent, xmlNode * new_obj)
-        equiv_node = pcmk__xe_first_child(parent, object_name, PCMK_XA_ID,
+    if (object_id != NULL) {
+        existing_node = pcmk__xe_first_child(parent, object_name, PCMK_XA_ID,
                                           object_id);
```
It'll have to be two ternaries, one for the attribute name and one for the value...

kgaillot commented on configure.ac:
```diff
@@ -1224,6 +1224,23 @@ AM_CONDITIONAL([WRAPPABLE_UNAME], [test x"$WRAPPABLE_UNAME" = x"yes"])
+                                        }
+                                   ]])],
+                  [],
+                  [AC_MSG_FAILURE("char must be 8 bits")])
```
being a little pedantic, I would say something like "Pacemaker does not support platforms where char is not 8 bits" (to acknowledge the burden is on us, not that the hardware is at fault)

nrwahl2 replied:
That is better

kgaillot commented on daemons/attrd/attrd_ipc.c:
```diff
@@ -269,7 +269,7 @@ expand_regexes(xmlNode *xml, const char *attr, const char *value, const char *re
                  * regex and replace it with the name.
                  */
-                attrd_copy_xml_attributes(xml, child);
+                pcmk__xe_copy_attrs(child, xml, false, false);
```
worth updating comment

nrwahl2 replied:
sounds good, will say "non-conflicting attributes", the rest still seems fine but let me know if you think of something else

kgaillot commented on lib/common/tests/xml/pcmk__xe_copy_attrs_test.c:
```diff
@@ -14,49 +14,51 @@
 
-    pcmk__xe_set_props(test_xml_1, "test_prop", "test_value", NULL);
+    crm_xml_add(src, "attr", "value");
+    pcmk__xe_copy_attrs(target, src, true, false);
```
overkill, but may as well test various combinations of bool args

nrwahl2 replied:
I'm redoing this with flags, and testing all combinations of flags is impractical in general. There will only be three for now but still.

kgaillot commented on lib/common/tests/xml/pcmk__xe_copy_attrs_test.c:
```diff
@@ -14,49 +14,51 @@
+    pcmk__xe_copy_attrs(target, src, true, false);
 
-    pcmk__xe_copy_attrs(test_xml_2, test_xml_1, true, false);
+    assert_ptr_equal(target, NULL);
```
This doesn't really test anything unfortunately. The call couldn't reassign target simply per C. Best approach would be to have `pcmk__xe_copy_attrs()` return EINVAL and test that.

nrwahl2 replied:
You're right. The previous version included the same test and I didn't notice while cleaning up.

kgaillot commented on lib/common/tests/xml/pcmk__xe_copy_attrs_test.c:
```diff
@@ -14,49 +14,51 @@
+    pcmk__xe_copy_attrs(target, src, true, false);
 
-    assert_ptr_equal(test_xml_2->properties, NULL);
+    assert_ptr_equal(target->properties, NULL);
```
This is slightly more useful but EINVAL would still be a good additional test

kgaillot commented on lib/common/tests/xml/pcmk__xe_copy_attrs_test.c:
```diff
@@ -14,49 +14,51 @@
+    xmlNode *target = pcmk__xe_create(NULL, "test");
 
-    pcmk__xe_copy_attrs(test_xml_2, test_xml_1, true, false);
+    pcmk__xe_copy_attrs(target, src, true, false);
```
ditto about bools

kgaillot commented on lib/common/schemas.c:
```diff
@@ -489,26 +489,33 @@ pcmk__sort_schemas(void)
+    static bool needs_init = true;
 
-    wrap_libxslt(false);
+    if (needs_init) {
```
Actually we rely on calling `crm_schema_init()` multiple times for the new remote schema stuff (see the call in `remoted_request_cib_schema_files()`). This will need to be a little more aware.

nrwahl2 replied:
Hmm, okay I'll look into that. I had one or more unit tests fail until I added this guard.

kgaillot replied:
I'm guessing it should be sufficient to add a force argument. If initialization has already been done, without force it would be a no-op, with force it would do a schema_cleanup and re-initialization. I haven't traced all callers to verify that though.

nrwahl2 replied:
Either that or have `crm_schema_cleanup()` set the `needs_init` flag back to false

kgaillot commented on lib/common/tests/schemas/pcmk__build_schema_xml_node_test.c:
```diff
@@ -33,6 +33,7 @@ static int
 setup(void **state) {
     setenv("PCMK_schema_directory", PCMK__TEST_SCHEMA_DIR, 1);
     crm_schema_init();
+    pcmk__xml_test_setup_group(state);
```
the explicit schema init is no longer needed

nrwahl2 replied:
right, which is why the test was failing (see comment in above thread). I figured I'd keep the explicit call to `crm_schema_init()` so that we don't have to worry about what is or is not in `pcmk__xml_test_setup_group()`, and the point of the `setenv()` is obvious. It is optional though

kgaillot commented on lib/common/tests/xml/pcmk__xe_copy_attrs_test.c:
```diff
@@ -58,7 +78,94 @@ copy_one(void **state)
+    xmlNode *src = pcmk__xe_create(NULL, "test");
+    xmlNode *target = pcmk__xe_create(NULL, "test");
+
+    // Overwrite true
```
separate tests for overwrite and do_not_overwrite would be cleaner (similar for score_update)

nrwahl2 commented on cts/cli/regression.acls.exp:
```diff
@@ -372,14 +372,9 @@ crm_attribute: Error performing operation: Permission denied
-pcmk__check_acl 	trace: User 'unknownguy' without ACLs denied read/write access to /cib/configuration/resources/primitive[@id]
-pcmk__apply_creation_acl 	trace: Creation of <primitive> scaffolding with id="<unset>" is implicitly allowed
+pcmk__check_acl 	trace: User 'unknownguy' without ACLs denied read/write access to /cib/configuration/resources/primitive[@id='dummy']
+pcmk__apply_creation_acl 	trace: ACLs disallow creation of <primitive> with id="dummy"
```
Thank you, I missed this in the sea of epoch changes.

I don't know yet but it's definitely worth looking into.

nrwahl2 replied:
From previous commit message:
> The only meaningful difference between this [update_cib_object()] and
> pcmk__xml_copy() is that this checks ACLs after copying a set of
> attributes and returns an error upon ACL denial. That error appears in
> the failed result XML from cib_process_create().
> 
> However, pcmk__xml_copy() still handles change tracking for ACL
> purposes. If there's an ACL denial, it will be caught later by
> cib_perform_op(), and the update will be discarded.
> 
> Catching a denial later is a solid tradeoff for the reduction in
> complexity IMO.

Previously, ACLs denied creation of attributes in the `<primitive>` element, and that caused an immediate error in `update_cib_object()`. Now, the error is caught only when ACLs are checked for the change as a whole later.

So we don't get a `<failed>` object because there's no longer a failure within `cib_process_create() -> add_cib_object()` like there was before.

Side note, `cibadmin` doesn't support `--output-as=xml` yet. This `<failed>` object only appears as undocumented, schema-less XML in text output.

> do you know why it happened?

been trying to figure out EXACTLY what differences exist...

kgaillot replied:
> So we don't get a `<failed>` object because there's no longer a failure within `cib_process_create() -> add_cib_object()` like there was before.
> 
> Side note, `cibadmin` doesn't support `--output-as=xml` yet. This `<failed>` object only appears as undocumented, schema-less XML in text output.

I believe it's also returned to the API callback (not that anything external is known to use the API). One question is why `update_results()` is called only in `cib_process_create()` and not more broadly.

nrwahl2 replied:
> I believe it's also returned to the API callback 

Yeah looks that way. I'd like to drop `<failed>` and `update_results()` altogether whenever that's allowed. It's used *only* with CIB create operatoins, its format isn't documented or codified in a schema, and it's mostly redundant.

> One question is why `update_results()` is called only in `cib_process_create()` and not more broadly.

It was created specifically for `cib_process_create()` in aa5e6743. Maybe it was planned to be used more extensively later and never was (or maybe it was, and those uses have since disappeared). Or maybe ACLs were implemented differently back then, so that we had to check for denials after adding attributes, but that seems less likely.

kgaillot replied:
> > I believe it's also returned to the API callback
> 
> Yeah looks that way. I'd like to drop `<failed>` and `update_results()` altogether whenever that's allowed. It's used _only_ with CIB create operatoins, its format isn't documented or codified in a schema, and it's mostly redundant.

Yeah that should be fine to deprecate for 2.1.8 and drop in 3.0.0. A deprecation message in the API doxygen should be sufficient.

> > One question is why `update_results()` is called only in `cib_process_create()` and not more broadly.
> 
> It was created specifically for `cib_process_create()` in [aa5e674](https://github.com/ClusterLabs/pacemaker/commit/aa5e6743f198aa90cb30a81e5b3a73593bdf3ab8). Maybe it was planned to be used more extensively later and never was (or maybe it was, and those uses have since disappeared). Or maybe ACLs were implemented differently back then, so that we had to check for denials after adding attributes, but that seems less likely.

nrwahl2 replied:
Lol I just noticed https://github.com/ClusterLabs/pacemaker/commit/aa5e6743f198aa90cb30a81e5b3a73593bdf3ab8#diff-43fdbe47b2127355408c6fbf9c07d8b8f2ab141eead9141f900ebf202084658cR1002

"remove this function" for `update_results()`

---

> It was created specifically for `cib_process_create()` in [aa5e674](https://github.com/ClusterLabs/pacemaker/commit/aa5e6743f198aa90cb30a81e5b3a73593bdf3ab8)

This was incorrect. Commit aa5e674 made a duplicate `update_cib_object()` function in a new file (cib_ops.c) without removing the old `update_cib_object()` in `primatives.c` (*sic*).

The original one goes all the way back to 1f442e5d0, where it *was* used in more places.

Mystery solved.

nrwahl2 replied:
> > do you know why it happened?

> been trying to figure out EXACTLY what differences exist...

Following up with more detail on the differing messages from `pcmk__check_acl()`:
* Previously, the denial happened when `update_cib_object()` -> `copy_in_properties()` -> `crm_xml_add()` called `pcmk__check_acl()`. We made it past the creation of the `<primitive>` element, and that was allowed during `pcmk__apply_creation_acl()`. However, the first attempt to set an attribute caused `crm_xml_add()` to error, which in turn caused `update_cib_object()` and `add_cib_object()` to error, leaving an empty `<primitive>` element.
* Now, we don't set any attributes explicitly; we copy the whole subtree. So we get our first error message from `pcmk__check_acl()` later, and it reflects the fact that `<primitive id="dummy">` was copied with its attributes. `cib_perform_op()` -> `xml_acl_disable()` -> `pcmk__apply_creation_acl()` (recursive) -> `pcmk__check_acl()`.

nrwahl2 commented on daemons/based/based_notify.c:
```diff
@@ -132,7 +132,7 @@ attach_cib_generation(xmlNode *msg)
 
     if (the_cib != NULL) {
-        pcmk__xe_copy_attrs(generation, the_cib);
+        pcmk__xe_copy_attrs(generation, the_cib, false);
```
That's true. This is also a good catch. `cibadmin.c` has the following in the help text:
```
           "Increase configuration version to prevent old configurations from "
               "being loaded accidentally:\n\n"
           "\t# cibadmin --modify --xml-text "
               "'<" PCMK_XE_CIB " " PCMK_XA_ADMIN_EPOCH
                   "=\"" PCMK_XA_ADMIN_EPOCH "++\"/>'\n\n"
```

Previously I was looking for `"++"` since looking at all `++` would've been infeasible.

I expected the `bump()` method to bump the admin epoch but in fact it bumps the epoch. Still, bumping the admin epoch seems like more of a use case for a method than for some obscure syntax.

kgaillot replied:
It would be for CIB updates, generally via cibadmin rather than C calls. A tool could build an update and put `++` in the admin_epoch.

kgaillot replied:
I'm leaning to keeping the functionality. It is useful, to avoid having to read and parse the CIB before submitting a simple change. And if we've been using it as an example all this time, it's more likely someone somewhere is using it. I'd be fine with deprecating `+=` but that's so trivial we may as well keep it too.

nrwahl2 replied:
The main reason I'd love to deprecate and drop it (unless I'm forgetting some other big reason) is the asymmetry. I don't think a replace operation should treat "value++" as a literal if a modify operation treats it as "value=<existing_value>+1".
* It makes behavior less consistent.
* It's harder to reason about, both due to the above inconsistency and because the XML attr value that we set is not the XML attr value that we get.
* It makes the code more difficult to maintain.

> It is useful, to avoid having to read and parse the CIB before submitting a simple change

Yep. IMO that doesn't outweigh the disadvantages of keeping it. Prior to this PR, it was used only once in internal code, and now zero times. It's not used on any performance-critical path. Having to do an extra query before incrementing the failcount during a simulation, or as a user before updating the admin_epoch, isn't a big deal. For the latter case, we could add a method.

> it's more likely someone somewhere is using it

Maybe. (Having worked support for 4.5 years, I wouldn't be surprised if no one is using that to be honest with you.) But that's why we'd deprecate it instead of dropping it immediately.

nrwahl2 replied:
> it's more likely someone somewhere is using it

also on this note, we never explicitly documented it AFAIK, only indirectly as part of that `admin_epoch` example. (that's a weak argument on its own but it's worth noting)

kgaillot replied:
I'm OK with deprecating it, but let's take it out of the example, log a deprecation warning when it's used, and keep it for the 3.0 series

It's not a big loss for interactive usage, but for automation it could be a serious pain. Consider an ansible role that modifies the CIB and bumps admin_epoch using `++`. It might be worth keeping support for just admin_epoch.

nrwahl2 replied:
Another option would be to have a "hidden" attribute, something like `__increment__="true"` or `pcmk__increment="true"`, that `cibadmin` can set. If true, expand anything like `value="value++"` in the input XML. Otherwise, treat all values literally. We don't necessarily have to publicly document it, although we could. (We can ignore `+=`.) It could apply to both replace and modify, whereas the current behavior only works for modify.

That way, the default behavior is intuitive, but there's still a way to request an atomic increment.

I'm reluctant to add another obscure attribute but it could give the best of both worlds.

nrwahl2 replied:
> I'm reluctant to add another obscure attribute but it could give the best of both worlds.

The only part this wouldn't resolve is "this sucks to maintain." But it would be comparable to what we have now, at least.

kgaillot replied:
> Another option would be to have a "hidden" attribute, something like `__increment__="true"` or `pcmk__increment="true"`, that `cibadmin` can set. If true, expand anything like `value="value++"` in the input XML. Otherwise, treat all values literally. We don't necessarily have to publicly document it, although we could. (We can ignore `+=`.) It could apply to both replace and modify, whereas the current behavior only works for modify.
> 
> That way, the default behavior is intuitive, but there's still a way to request an atomic increment.
> 
> I'm reluctant to add another obscure attribute but it could give the best of both worlds.

We wouldn't need a special attribute; we could add a CIB call option.

nrwahl2 replied:
Oh you're right.

I lean toward using simply "++" instead of "name++" as the value for the new way. Simplifies parsing. As for "+=", it would be about as easy to keep it as it is to drop it. I guess keep it. Otherwise if it turns out we need it, adding it would be a breaking change.

kgaillot replied:
Agreed, but we should keep `name++` so that any existing scripts don't break. If only cibadmin uses the new option, it makes everything else more efficient while preserving the one reasonable use case. The CIB digests prevent simple editing of the file, so everyone really does have to go through cibadmin. I would also use it in crm_attribute's update option, which can go directly through the CIB API. I don't think any other tools would benefit from it.

nrwahl2 replied:
> we should keep `name++` so that any existing scripts don't break

The way I envisioned it, existing scripts would break anyway at a compatibility break. I figured we'd add a CLI option for the increment behavior. That way the user can specify that they want `"++"`/`"+=X"` to expand to "existing value plus one/plus X". Otherwise, no attempt is made to expand values, and what you input is what you get.

(Suppose that for some strange reason a user wants to set `"++"` as the literal value; they should be able to do so without any surprises. If they want special behavior, they can set a flag requesting it.)

Until some compatibility break, also preserve the existing "expand `value++` to `<value + 1>` by default for modify ops" behavior. But replace the example in the help output, with the new syntax that includes the CLI flag.

nrwahl2 replied:
planning to demo it as soon as I'm able and you can see if you like it or not. Classes today so might not be until tomorrow

kgaillot replied:
I wouldn't bother with a new option or syntax, just have cibadmin support the current syntax

nrwahl2 replied:
That's easy but leaves us in the same boat as now, where modify and replace behave differently (making replace do it automatically would be a breaking change), and where certain values are treated as "special" and transformed without it being requested

It's gonna have to support the current syntax until a compat break regardless, I mean going forward though

Realistically, the impact is minimal whatever we decide. I'm only coming at this from a "principle of least astonishment" POV. To me the least astonishing scenario is that I can set whatever value I want and it's treated
* literally
* the same regardless of how I request it

kgaillot replied:
I'm mainly concerned with the possibility of existing users depending on it, whether for admin_epoch via cibadmin or a permanent node attribute via crm_attribute. Both uses are actually reasonable, so I think it's worth keeping support. I'm thinking:

Keep the current syntax, forever, because it's not a big deal and it's one less thing existing users would have to change.

Add two new CIB call options that both enable the behavior. One logs a deprecation warning when the syntax is used.

Add command-line options to cibadmin and crm_attribute to enable the behavior. Show the new option in the cibadmin example. If the option is specified, they should pass the new call option without the warning; if the option is not specified, they should pass the new call option with the warning. The end result is that the behavior is still always supported, but the deprecation warning is logged if the new command-line option is not used.

Drop support for the deprecated usage in 3.1.0 or later.

nrwahl2 replied:
This seems like a good approach

nrwahl2 replied:
It is a *little* bit awkward that we treat these updates as scores, since one of the main anticipated use cases is for bumping the `admin_epoch`. Which is definitely not a score. Other use cases may also be regular old integer counters. I don't want to open that can of worms though.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -460,6 +460,73 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
+        const char *v = value;
+
+        // Stop at first character that differs between name and value
+        for (; (*n == *v) && (*n != '\0'); n++, v++);
```
If they're not the same length, then we'll exit when `(*n != *v)` because one of them isn't `'\0'`. The `'\0'` check is really checking for the case when they're both `'\0'`

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -460,6 +460,73 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
+{
+    const char *old_value = NULL;
+
+    CRM_CHECK((target != NULL) && (name != NULL) && (value != NULL), return);
```
ouch, good catch. Even if existing callers don't, it's needed for parity with `crm_xml_add()`.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -460,6 +460,73 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
+ *
+ * \note The original attribute value in \p target and the number in an
+ *       assignment expression in \p value are parsed as scores. For more
+ *       details, refer to \c char2score().
```
Good point thanks

nrwahl2 commented on cts/cli/regression.validity.exp:
```diff
@@ -473,7 +473,7 @@ element rsc_order: Relax-NG validity error : Element constraints has extra conte
 element rsc_order: Relax-NG validity error : Element constraints has extra content: rsc_order
 =#=#=#= Current cib after: Make resulting CIB invalid, and without validate-with attribute =#=#=#=
-<cib epoch="41" num_updates="0" admin_epoch="0" validate-with="none">
+<cib epoch="11" num_updates="0" admin_epoch="0" validate-with="none">
```
If I'd known that the straightforward changes to `update_cib_object()` were going to turn into such an ordeal, I wouldn't have made them. But here we are...

I tested using the first cts-cli validity test that changed: https://github.com/ClusterLabs/pacemaker/pull/3403/commits/8d49c56cc392ad16e7b73dad6bf42fcd0654fd8e#diff-711f8daed23fbf12c53cc6813a6d50307dc79787757201c5105f15877170803aL5

For now I'll assume this change has the same cause. Previously, `xml_create_patchset()` correctly viewed newly created elements within `<configuration>` as a config change. Now, it does not. The reason? `pcmk__xml_copy()` doesn't mark the parents or document dirty. It only calls `pcmk__mark_xml_created()` on the copy.

`pcmk__mark_xml_created()` calls `pcmk__mark_xml_node_dirty()` only if the `pcmk__xf_created` flag is NOT already set. But `pcmk__xf_created` IS already set, because that's part of the `new_private_data()` function that gets called when any XML node is created.
```
pcmk__mark_xml_created:
    if (nodepriv && pcmk__tracking_xml_changes(xml, FALSE)) {
        if (!pcmk_is_set(nodepriv->flags, pcmk__xf_created)) {
            pcmk__set_xml_flags(nodepriv, pcmk__xf_created);
            pcmk__mark_xml_node_dirty(xml);
        }
        ...
    }
```

This has far-reaching implications beyond what changed in this commit. It means that for both `pcmk__xe_create()` and `pcmk__xml_copy()`, if a new node or copy is added to an existing node, the chain of parents and the document do NOT get marked dirty. So if the change is in `<configuration>`, it's not detected as a config change, and the epoch doesn't get bumped.

This probably long-standing behavior looks like a bug.

This should be easy to fix (probably within `pcmk__mark_xml_created()` by moving the `pcmk__mark_xml_node_dirty()` call outside the innermost `if`). I hope it doesn't break other code that was built around the buggy behavior.

kgaillot replied:
If that's the case, we can simply drop the `if`

kgaillot replied:
The `if` was added in 5125ecbdfe, but it seems unrelated to the stated purpose of the commit, and I don't see any reason for it. I suspect it was a mistake.

nrwahl2 replied:
> drop the `if`

yep that's what I ended up doing :)

> I don't see any reason for it

Maybe
* Avoid logging for setting a bit that's already set -- but `pcmk__set_xml_flags()` uses `LOG_NEVER` so it's a no-op aside from `|=`.
* Avoid wasteful calls to `pcmk__mark_xml_node_dirty()`. Note that for every descendant node, `pcmk__mark_xml_node_dirty()` would trace back through the entire parent chain and doc. But that's avoidable and I've avoided it in my changed commit.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -460,6 +460,31 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
+void
+pcmk__xe_copy_attrs(xmlNode *target, const xmlNode *src)
+{
+    CRM_CHECK((src != NULL) && (target != NULL), return);
```
> Have you double-checked that callers can never pass NULL?

Effectively yes. The one caveat is that `cib_perform_op()`'s checking is not great. It null-checks `current_cib` but not `*current_cib`, except that in some cases it will try to dereference `*current_cib`...

`cib_perform_op()` calls `pcmk__xe_copy_attrs()` both directly and via `cib_process_erase()`, using `*current_cib`.

nrwahl2 replied:
> xml_element.c

https://projects.clusterlabs.org/T815

nrwahl2 commented:
The most recent push addresses some but not all of review. In particular, it addresses the epoch bump behavior discussed in https://github.com/ClusterLabs/pacemaker/pull/3403#discussion_r1558857982.

nrwahl2 commented:
rebased on main

kgaillot commented:
changes so far look good

nrwahl2 commented:
> cts-cli is failing on FreeBSD, can you take a look? Might be unrelated to your changes
> 
> ```
> [2024-04-10T16:00:12.941Z] * Running: cibadmin - niceguy: Replace - create resource
> [2024-04-10T16:00:13.505Z] * Failed (rc=138): cibadmin       - niceguy: Replace - create resource (./tools/cibadmin)
> ```

It looks like it ought to be, but my changes passed CI fine. Weird.

I'm unable to access the CI machines right now. Looks like maybe a DNS issue; I'm waiting to hear back from Fabio. In the meantime, here's the test output diff from the logs, showing core dumps:
```
06:58:30  Results
06:58:30  --- ./cts/cli/regression.acls.exp	2024-04-11 14:26:35.000000000 +0000
06:58:30  +++ /tmp/cts-cli.acls.8B9CcAp3BE	2024-04-11 14:58:18.928151000 +0000
06:58:30  @@ -1187,9 +1187,9 @@
06:58:30   pcmk__check_acl 	trace: Default ACL denies user 'niceguy' read/write access to /cib[@epoch]
06:58:30   pcmk__check_acl 	trace: Default ACL denies user 'niceguy' read/write access to /cib/configuration/resources/primitive[@id='dummy2']
06:58:30   pcmk__apply_creation_acl 	trace: ACLs disallow creation of <primitive> with id="dummy2"
06:58:30  -Call failed: Permission denied
06:58:30  -=#=#=#= End test: niceguy: Replace - create resource - Insufficient privileges (4) =#=#=#=
06:58:30  -* Passed: cibadmin       - niceguy: Replace - create resource
06:58:30  +./cts/cts-cli: line 152: 36510 Bus error               (core dumped) cibadmin --replace --xml-file /tmp/cts-cli.acls.xml.WA5JZIgkKd
06:58:30  +=#=#=#= End test: niceguy: Replace - create resource - Interrupted by signal (138) =#=#=#=
06:58:30  +* Failed (rc=138): cibadmin       - niceguy: Replace - create resource
06:58:30   <cib epoch="14" num_updates="0" admin_epoch="0">
06:58:30     <configuration>
06:58:30       <crm_config>
06:58:30  @@ -3228,9 +3228,9 @@
06:58:30   pcmk__check_acl 	trace: Default ACL denies user 'niceguy' read/write access to /cib[@epoch]
06:58:30   pcmk__check_acl 	trace: Default ACL denies user 'niceguy' read/write access to /cib/configuration/resources/primitive[@id='dummy2']
06:58:30   pcmk__apply_creation_acl 	trace: ACLs disallow creation of <primitive> with id="dummy2"
06:58:30  -Call failed: Permission denied
06:58:30  -=#=#=#= End test: niceguy: Replace - create resource - Insufficient privileges (4) =#=#=#=
06:58:30  -* Passed: cibadmin       - niceguy: Replace - create resource
06:58:30  +./cts/cts-cli: line 152: 38688 Bus error               (core dumped) cibadmin --replace --xml-file /tmp/cts-cli.acls.xml.WA5JZIgkKd
06:58:30  +=#=#=#= End test: niceguy: Replace - create resource - Interrupted by signal (138) =#=#=#=
06:58:30  +* Failed (rc=138): cibadmin       - niceguy: Replace - create resource
```

nrwahl2 commented:
rebased on main

nrwahl2 commented:
Upcoming push is incomplete. In particular:
* The unit tests haven't been updated yet
* The `pcmk__xa_copy_score_update_warn` flag hasn't been plugged in yet. We should warn within `pcmk__xe_set_score()` and only if we find the "++"/"+=" syntax in combination with the warn flag. If we warn just because the flag is set, then we'll warn every time `cibadmin` or `crm_attribute` is run.
* The `cib_score_update_warn` flag hasn't been added yet. It will need to be set by default in `cibadmin` and `crm_attribute` if the new `--score-update` (or whatever) CLI option was not used.
* The new `cibadmin` and `crm_attribute` CLI options have not been added. This depends on the preceding tasks.

However, since you'll be gone by the time I wake up today, I'm pushing what I have. If you see any problems or want something renamed, let me know.

nrwahl2 commented:
Pushed again, I think all review is addressed. I haven't looked back over the whole thing since finishing, so there may be errors. I plan to do that today but you might beat me to it. It does compile though, and tests pass on my machine.

One note: The warning from `cibadmin` and `crm_attribute` (via `crm_warn()` in `pcmk__xe_set_score()`) is basically useless, because it appears only when a few `-V` flags are set. Off hand, I'm not sure of a good way around that. One idea is to create a new standard Pacemaker return code for this purpose, to signal that an attribute value was expanded via the deprecated `cib_score_update_warn` flag.

It's not great form to use an error code to indicate a warning, but hey.

kgaillot commented on include/crm/cib/cib_types.h:
```diff
@@ -129,6 +129,20 @@ enum cib_call_options {
+     * \deprecated For internal use only. This value will be removed in a future
+     * release.
+     */
+    cib_score_update_warn   = (1 << 15),
```
similarly this would be unnecessary if we can pass a flag back to the client

kgaillot commented on tools/cibadmin.c:
```diff
@@ -388,6 +389,33 @@ static GOptionEntry addl_entries[] = {
       INDENT "Default value: 'auto'",
       "[value]" },
 
+    { "score-update", 0, G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE,
```
just "score" should be fine

nrwahl2 replied:
That is more concise and the description goes into detail, so I'm okay with it. One counterpoint: the value is not treated as a score in general. `attr="1000001"` will set `1000001`, not `1000000` (INFINITY).

kgaillot replied:
Yeah that is unfortunate but the description should be sufficient

kgaillot commented on tools/crm_attribute.c:
```diff
@@ -298,6 +299,34 @@ static GOptionEntry addl_entries[] = {
       "SECTION"
     },
 
+    { "score-update", 0, G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE,
```
should default to true (even in the future) if --promotion is given

kgaillot commented on include/crm/common/xml_internal.h:
```diff
@@ -425,6 +425,19 @@ xmlNode *pcmk__xe_next_same(const xmlNode *node);
+
+    //! Don't overwrite existing values
+    pcmk__xa_copy_no_overwrite  = (1U << 0),
+};
```
Since this is likely to see more use, s/xa_copy/xaf/

nrwahl2 replied:
They're not flags for attributes (which sort of suggests that they apply to the attrs themselves), they're specifically flags for copying/updating attributes. The most general we could say is that they're flags for attr operations.

`xaf` makes the lengths more manageable so I'm not gonna fight about it either way.

nrwahl2 replied:
could do something like `xa_op_*`... it's internal and I'm not that worried about current choice, but `xaf` does come out very close to `xml_private_flags`: `pcmk__xf_*`.

going with `xaf` for now, just a thought

kgaillot replied:
maybe `pcmk__xav_*` (for XML attribute value)

kgaillot reviewed (commented):
nice

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -551,8 +553,19 @@ pcmk__xe_set_score(xmlNode *target, const char *name, const char *value)
+                         "deprecated unless cib_score_update flag is set)",
+                         name, value, name, new_value_i, target->name,
+                         ((id != NULL)? " " PCMK_XA_ID "=" :  ""),
+                         pcmk__s(id, ""));
```
This is what I was getting at with:

> One note: The warning from `cibadmin` and `crm_attribute` (via `crm_warn()` in `pcmk__xe_set_score()`) is basically useless, because it appears only when a few `-V` flags are set. Off hand, I'm not sure of a good way around that. One idea is to create a new standard Pacemaker return code for this purpose, to signal that an attribute value was expanded via the deprecated `cib_score_update_warn` flag.
> 
> It's not great form to use an error code to indicate a warning, but hey.

(https://github.com/ClusterLabs/pacemaker/pull/3403#issuecomment-2068820950)

kgaillot replied:
It wouldn't be indicating a warning, it would be indicating "success with expansion". The client would choose to make that a warning.

nrwahl2 commented:
@kgaillot This isn't completely finished. I pushed the current state so you can let me know of any concerns.

Using a `pcmk_rc_expanded_score` return code seems ideal as it allows the tools to use their output objects to emit a warning. However, that return code has to bubble up through the call chain, which has turned out to be a pain.
* Existing handling of legacy RCs is inconsistent and sometimes doesn't consider sync calls.
* I put some some ugly code in several places to do things like replace `rc` with `local_rc` (from some update call) if it's `pcmk_rc_expanded_score` and otherwise treat `pcmk_rc_expanded_score` the same as `pcmk_rc_ok`. Don't want to lose that extra info if we find it. May be worth functionizing but it's short and fairly special-purpose...

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -2323,23 +2323,33 @@ pcmk__xe_foreach_child(xmlNode *xml, const char *child_element_name,
-                return rc;
+            int local_rc = handler(node, userdata);
+
+            if (local_rc == pcmk_rc_expanded_score) {
```
case should be added to unit tests if we keep this

nrwahl2 commented:
FWIW we won't strictly need to keep the `pcmk_rc_expanded_score` return code at all after we drop the "score update and warn by default" behavior from tools

kgaillot commented:
Basic idea looks good. I didn't check whether existing callers will handle the new rc correctly.

> FWIW we won't strictly need to keep the `pcmk_rc_expanded_score` return code at all after we drop the "score update and warn by default" behavior from tools

probably worth removing at that point for simplicity

nrwahl2 commented:
> probably worth removing at that point for simplicity

Nice. I think we ought to add it with a `\deprecated` flag and note in the first place then.

nrwahl2 commented:
only rebased on main

nrwahl2 commented on tools/cibadmin.c:
```diff
@@ -775,9 +775,13 @@ main(int argc, char **argv)
-    if (rc > 0) {
-        /* wait for the reply by creating a mainloop and running it until
-         * the callbacks are invoked...
+    if (!pcmk_is_set(options.cmd_options, cib_sync_call)
```
Soon I'd like to make all CIB calls synchronous in `cibadmin`, unless we come up with a reason not to. Deprecate `--sync-call`. Not sure why we have these two separate code paths when the user running this from the command line has to wait for the call to receive a result either way.

kgaillot replied:
I believe it was originally for testing the code paths. Not needed anymore

nrwahl2 replied:
https://projects.clusterlabs.org/T817

nrwahl2 commented:
I've got a "best-effort" warning system implemented. It's plugged in only where it needs to be, based on where we pass `cib_return_expanded_score` for backward compatibility. It's also not unit-tested. As noted in the comments, I think worst-case scenario is that we return false and miss a warning.

nrwahl2 commented:
note to self, forgot to update the example in `cibadmin`

kgaillot reviewed (commented):
I don't think the warning is worth the potential for bugs from all the changes needed, or the efficiency hit from comparing entire trees. Let's just kill it.

nrwahl2 commented:
> I don't think the warning is worth the potential for bugs from all the changes needed, or the efficiency hit from comparing entire trees. Let's just kill it.

I don't think the efficiency hit is relevant, since this only comes up in CLI tool invocations.

I'm fine with killing it. I was so close when I told you about the difficulties, that I wanted to give it a proper shot at completion...

nrwahl2 commented:
Dropped the commits for implementing the warning, and edited the deprecation commits as necessary
